File: python/problems/basic-calculator-iii.py
from local_stuff import *


class Reader:
  def __init__(self, s: str):
    self.value = s
    self.index = 0

  def read_num(self):
    r = 0
    while self.index < len(self.value) and self.value[self.index].isdigit():
      r = r * 10 + (ord(self.value[self.index]) - ord('0'))
      self.index += 1

    return r

  def is_num(self):
    return self.value[self.index].isdigit()

  def read_char(self):
    self.index += 1
    return self.value[self.index - 1]

  def peek(self):
    return self.value[self.index]

  def has_next(self) -> bool:
    return self.index < len(self.value)

class Solution:
  def eval_stk(self, stk):
    res = stk[0]
    for i in range(1, len(stk), 2):
      if stk[i] == '+':
        res += stk[i + 1]
      else:
        res -= stk[i + 1]

    return res

  def solve(self, reader: Reader) -> int:
    stk = []

    while reader.has_next():
      c = reader.peek()
      if c == ')':
        reader.read_char()
        return self.eval_stk(stk)

      if c == ' ':
        reader.read_char()
        continue

      if c in ['*', '/', '+', '-']:
        stk.append(reader.read_char())
      else:
        if c == '(': reader.read_char()
        b = self.solve(reader) if c == '(' else reader.read_num()

        if stk and (stk[-1] == '/' or stk[-1] == '*'):
          op = stk.pop()
          a = stk.pop()
          stk.append((a * b) if op == '*' else int(a / b))
        else:
          stk.append(b)

    return self.eval_stk(stk)


  def calculate(self, s: str) -> int:
    return self.solve(Reader(s))

File: cpp/leetcode/problems/basic-calculator-iii.cpp
#ifdef RUNNING_LOCALLY
#include "local-stuff.hpp"
#endif

class Reader {
 public:
  int index;
  string value;
  Reader(string& s) : index(0), value(s) {}

  int read_num() {
    int r = 0;
    while (index < value.length() && isdigit(value[index])) {
      r = r * 10 + (value[index++] - '0');
    }
    return r;
  }

  char is_num() { return isdigit(value[index]); }
  char read_char() { return value[index++]; }
  char peek() { return value[index]; }
  bool has_next() { return index < value.length(); }
};

class Solution {
 public:
  int eval_stk(vector<variant<int, char>>& stk) {
    int res = get<int>(stk[0]);
    for (int i = 1; i < stk.size(); i += 2) {
      if (get<char>(stk[i]) == '+') {
        res += get<int>(stk[i + 1]);
      } else {
        res -= get<int>(stk[i + 1]);
      }
    }

    return res;
  }

  int solve(Reader& reader) {
    vector<variant<int, char>> stk;

    while (reader.has_next()) {
      char c = reader.peek();
      if (c == ')') {
        reader.read_char();
        return eval_stk(stk);
      }

      if (c == ' ') {
        reader.read_char();
        continue;
      }

      if (c == '*' || c == '/' || c == '+' || c == '-') {
        stk.push_back(reader.read_char());
      } else {
        if (c == '(') reader.read_char();
        int num = c == '(' ? solve(reader) : reader.read_num();

        if (!stk.empty() && (get<char>(stk.back()) == '/' || get<char>(stk.back()) == '*')) {
          int b = num;
          auto op = get<char>(stk.back());
          stk.pop_back();
          int a = get<int>(stk.back());
          stk.pop_back();
          stk.push_back(op == '*' ? (a * b) : (a / b));
        } else {
          stk.push_back(num);
        }
      }
    }

    return eval_stk(stk);
  }

  int calculate(string s) {
    Reader reader{s};
    return solve(reader);
  }
};

#ifdef RUNNING_LOCALLY
int main() {
  utils::Input in("../input.txt");
  while (in.has_next() && !in.peek().empty()) {
    Solution s;
    auto a = in.s();
    auto result = s.calculate(a);
    utils::expect_equal(result, in);
  }
}
#endif
